### ISSUE TITLE
[iOS] Deeplink + Confirmation Prompt issues

### ISSUE THREAD

**Issue by darthmeme** [MetaMask/metamask-mobile#4827]
**Describe the bug**
When any action requires the user to approve or reject said action (like `sendTransaction`) in MetaMask, more often than not the deeplink will re-direct to the AppStore instead of MetaMask even when it's installed. In addition, when finally opening MetaMask, half the time the confirmation prompt will never show.

**To Reproduce**
Requirements: MetaMask mobile (signed in)

1. In Safari, go to https://web3modal.com/ and click on "Connect". Choose WalletConnect, and then MetaMask. Finally, approve the connection.
2. Once redirected back to the web app, click on `eth.sendTransaction`

*Note*: I used web3modal here cause it's easiest to reproduce, but I've reproduced this issue on my own app as well as other web apps such as OpenSea.

**Expected behavior**
I expect the MetaMask app to open and the confirmation prompt for the transaction to show every time.

Instead I very often get redirected to the AppStore. And when I open MetaMask, the confirmation prompt doesn't show most of time. Although sometimes it does.

**Smartphone (please complete the following information):**
 - Device: iPhone 12 Pro
 - OS: iOS 15.6
 - App Version: 5.4.0 (931)

-------------------------------------------------------------
_to be added after bug submission by internal support / PM_
**Severity**

- How critical is the impact of this bug on a user?
- Add stats if available on % of customers impacted
- Is this visible to all users?
- Is this tech debt?

**Comment by githubisagit** [comment]
Same issue here, thought this was fixed in version 5.2. Game of chance as to whether MetaMask will show a confirmation prompt

**Comment by matthewrkula** [comment]
We are seeing a similar thing on Android. Multiple devices but my main test phone is a Pixel 6. We get inconsistent prompting for WalletConnect connections and sending transactions. It fails to show the proper prompt around 70% of the time which makes the transactions borderline unusable.

**Comment by cortisiko** [comment]
hey @darthmeme can you try to reproduce on v5.7.0? Thanks!

**Comment by darthmeme** [comment]
@cortisiko - Hey! Just tested and it's still not working. Seems worse this time round, now when I try and connect my wallet - the deep link will, without fail, bring me to the App Store. Then once I open MetaMask, the connection prompt never shows.

**Comment by chrisleewilcox** [comment]
@andreahaku please review.  Please see @tommasini.

**Comment by darthmeme** [comment]
Hi,

It's been a few months, are there any updates on this?

I've found that the App Store redirect issue are caused by [this](https://docs.walletconnect.com/2.0/swift/guides/mobile-linking#ios-app-link-constraints).

> The wallet interaction must be triggered as soon as possible within the event handler. Any preceding asynchronous work (e.g. estimating gas, resolving an ENS name, fetching a nonce) should have already completed before the event handler fires. This may require you to design the user experience around this constraint, preventing users from initiating a wallet interaction until it's ready rather than doing the work lazily.

Making sure the transaction data is ready to send immediately when the user initiates some action always redirects to MetaMask.

The only time it doesn't is when (at least on iOS), you toggle low battery mode. You'll be asked whether you want to open the MetaMask app. If you click "cancel" then from that moment on, it will always redirect to the App Store. The only way I've found to "reset" this behavior is to clearing all cache.

Unfortunately, there still seems to be a big issue where MetaMask will open but the prompt confirmation will never show, especially if the app is not running in the background (cold start).
